pig/src/org/partiql/pig/domain/include/IncludeResolver.kt

```diff
@@ -0,0 +1,100 @@
+ *
+ * @throws InvalidIncludePathException if one of [searchDirs] does not exist or is not a directory.
+ */
+internal class IncludeResolver(
+    searchDirs: List<Path>,
+    internal val fileSystem: FileSystem
+) {
+
+    private val searchDirectories = searchDirs
+        .map { it.toAbsolutePath().normalize() }
+        .onEach {
+            validatePath(it)
+            if (!Files.isDirectory(it)) {
+                throw InvalidIncludePathException(it.toString())
+            }
+        }.toTypedArray()
+
+    private fun validatePath(it: Path) {
+        require(it.fileSystem === fileSystem) { "Path $it must not belong to a different FileSystem instance." }
+    }
+
+    /**
+     * Searches for the absolute path of an included file, returning the first file found.
+     *
+     * The parent directory of the [includerFile] is searched first, followed by each of the source roots in turn. If
+     * [fileToInclude] starts with `/`, the parent directory of [includerFile] is skipped.
+     *
+     * @return The absolute [Path] of the first file located.
+     */
+    fun resolve(includeeFile: Path, includerFile: Path): Path {
+        val normalizedIncluder = includerFile.normalize().toAbsolutePath()
+
+        // Determine list of all possible search roots, excluding the includer's parent directory if
+        // the includee an absolute path.  T
+        val includerParentDir = normalizedIncluder.parent
+        val searchPaths = listOfNotNull(
+            includerParentDir.takeIf { !includeeFile.isAbsolute },
+            *searchDirectories
+        ).distinct()
+        // distinct is needed because duplicate entries can arise in this list, for instance if the
+        // includer's parent directory is the same as a include directory. Primarily this is needed to
+        // prevent the directory from appearing twice in the error messages we display when we can't
+        // find an include file.
+
+        val possibleIncludeeFiles = searchPaths
+            .map {
+                // Truncates / at the beginning of an absolute path This forces absolute paths absolute of includees
+                // to be relative the include paths specified on the command-line and prevents them from being relative
+                // to the includer.
+                val appendPath = when {
+                    includeeFile.isAbsolute -> includeeFile.toString().substring(1)
+                    else -> includeeFile.toString()
+                }
+                it.resolve(appendPath)
+            }
+
+
+        // The resolved file is the first one that exists.
+        return possibleIncludeeFiles.firstOrNull { Files.exists(it) }
+            // TODO: can we used a algebraic type here instead of an exception?
+            ?: throw IncludeResolutionException(
+                inputFilePathString = includeeFile.toString(),
+                consideredFilePaths = possibleIncludeeFiles.map { "${it}" })
```

Comment:
nit: can just use `"$it"`
```suggestion
                consideredFilePaths = possibleIncludeeFiles.map { "$it" })
```